# GeneMiner
# introduction
	GeneMiner is a software for extracting phylogenetic markers from second-generation sequencing (NGS) data.
	GeneMiner uses a K-mer filtering method to obtain reliable reads, then a weighted seed extension algorithm based on de Bruijn graph method to mine target sequences by assembling small target regions of short-read sequencing datasets, and finally the results are verified by an iterative method based on the   base substitution model.
	With GeneMiner, users can accurately and efficiently obtain a large number of target sequences from a wide range of histological data at a very low cost, such as: all or part of the chloroplast/mitochondrial genome, highly repetitive regions in the nuclear genome (e.g. nrDNA), etc. from shallow whole-genome sequencing data; single-low copy genes from transcriptome sequencing data, etc.  GeneMiner improves existing phylogenetic research strategies at the most basic data level, and has significant advantages in reducing experimental costs and expanding phylogenetic marker selection. In addition, GeneMiner can be applied to the extraction of DNA barcodes, customs inspection and quarantine, mining of specific functional genes and other research, which has broad application prospects.
